internal/clusterfeature/features/middleware.go

```diff
@@ -0,0 +1,51 @@
+// Copyright © 2019 Banzai Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package features
+
+import (
+	"net/http"
+
+	"emperror.dev/emperror"
+	"github.com/banzaicloud/pipeline/cluster"
+	ginutils "github.com/banzaicloud/pipeline/internal/platform/gin/utils"
+	"github.com/banzaicloud/pipeline/pkg/ctxutil"
+	"github.com/gin-gonic/gin"
+	"github.com/moogar0880/problems"
+)
+
+// NewMiddleware returns a new gin middleware that checks cluster is exists in the current org.
+func NewMiddleware(manager *cluster.Manager, errorHandler emperror.Handler) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		clusterID, ok := ginutils.UintParam(c, "id")
+		if !ok {
+			return
+		}
+
+		orgID, ok := ginutils.UintParam(c, "orgid")
+		if !ok {
+			return
+		}
+
+		_, err := manager.GetClusterByID(c, orgID, clusterID)
+		if err != nil {
+			errorHandler.Handle(err)
+			problem := problems.NewDetailedProblem(http.StatusNotFound, err.Error())
```

Comment:
The manager might respond with error for various reasons (eg. DB unavailable). Check if the returned error is a not found error (the root cause implements`interface{NotFound() bool}` and it returns true, see other examples in the repo) and return 404 only if that's the case.